### PR TITLE
add replication metrics

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

This PR adds several new metrics:

* `arangodb_network_forwarded_requests` to track the number of requests forwarded from one coordinator to another in a load-balancing context.

* `arangodb_aql_query_time`: histogram with AQL query times distribution.
* `arangodb_aql_slow_query_time`: histogram with AQL slow query times distribution.
* `arangodb_aql_slow_query`: total number of slow AQL queries.

* `arangodb_replication_dump_requests`: number of replication dump requests made.
* `arangodb_replication_dump_bytes_received`: number of bytes received in replication dump requests.
* `arangodb_replication_dump_documents`: number of documents received in replication dump requests.
* `arangodb_replication_dump_request_time`: wait time for replication dump requests.
* `arangodb_replication_dump_apply_time`: time required for applying data from replication dump responses.
* `arangodb_replication_initial_sync_keys_requests`: number of replication initial sync keys requests made.
* `arangodb_replication_initial_sync_docs_requests`: number of replication initial sync docs requests made.
* `arangodb_replication_initial_sync_docs_requested`: number of documents requested via replication initial sync requests.
* `arangodb_replication_initial_sync_docs_inserted`: number of documents inserted by replication initial sync.
* `arangodb_replication_initial_sync_docs_removed`: number of documents inserted by replication initial sync.
* `arangodb_replication_initial_chunks_requests_time`: wait time for replication key chunks determination requests.
* `arangodb_replication_initial_keys_requests_time`: wait time for replication keys requests.
* `arangodb_replication_initial_docs_requests_time`: time needed to apply replication docs data.
* `arangodb_replication_initial_insert_apply_time`: time needed to apply replication initial sync insertions.
* `arangodb_replication_initial_remove_apply_time`: time needed to apply replication initial sync removals.
* `arangodb_replication_initial_lookup_time`: time needed for replication initial sync key lookups.
* `arangodb_replication_tailing_requests`: number of replication tailing requests.
* `arangodb_replication_tailing_follow_tick_failures`: number of replication tailing failures due to missing tick on leader.
* `arangodb_replication_tailing_markers`: number of replication tailing markers processed.
* `arangodb_replication_tailing_documents`: number of replication tailing document inserts/replaces processed.
* `arangodb_replication_tailing_removals`: number of replication tailing document removals processed.
* `arangodb_replication_tailing_bytes_received`: number of bytes received for replication tailing requests.
* `arangodb_replication_tailing_request_time`: wait time for replication tailing requests.
* `arangodb_replication_tailing_apply_time`: time needed to apply replication tailing markers.

- [ ] :hankey: Bugfix 
- [x] :pizza: New feature 
- [ ] :hammer: Refactoring 
- [x] :book: CHANGELOG entry made
- [x] :muscle: The behavior in this PR was *manually tested*
- [ ] :computer: The behavior change can be verified via automatic tests

#### Backports:

- [x] No backports required
- [ ] Backports required for: 

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/536
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

- [x] This change is a trivial rework / code cleanup without any test coverage.

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11671/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11683/

File: CHANGELOG [hunk 1]
```diff
@@ -1,9 +1,69 @@
...
-  This can help debugging and inspecting the follower.
+* Added new metric `arangodb_network_forwarded_requests` to track the number
+  of requests forwarded from one coordinator to another in a load-balancing
+  context.
+
+* Added new metrics for tracking AQL queries and slow queries:
+  * `arangodb_aql_query_time`: histogram with AQL query times distribution.
+  * `arangodb_aql_slow_query_time`: histogram with AQL slow query times 
+    distribution.
+  * `arangodb_aql_slow_query`: total number of slow AQL queries.
```
Comment (jsteemann) [inline CHANGELOG:12]:
```suggestion
  * `arangodb_aql_all_query`: total number of all AQL queries.
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11690/

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 2]
```diff
@@ -824,7 +744,12 @@ Result DatabaseInitialSyncer::fetchCollectionDump(arangodb::LogicalCollection* c
 
   std::string const typeString = (coll->type() == TRI_COL_TYPE_EDGE ? "edge" : "document");
 
-  InitialSyncerDumpStats stats;
+  // statistics which will update the global replication metrics,
+  // periodically resetted 
+  ReplicationMetricsFeature::InitialSyncStats stats(vocbase().server().getFeature<ReplicationMetricsFeature>(), true);
+
+  // local statistics that will be ever increasing inside this method
+  ReplicationMetricsFeature::InitialSyncStats cumulatedStats(vocbase().server().getFeature<ReplicationMetricsFeature>(), false);
```
Comment (danielhlarkin) [inline arangod/Replication/DatabaseInitialSyncer.cpp:752]:
```suggestion
  // statistics which will update the global replication metrics,
  // periodically reset
  ReplicationMetricsFeature::InitialSyncStats stats(vocbase().server().getFeature<ReplicationMetricsFeature>(), true);

  // local statistics that will be ever increasing inside this method
  ReplicationMetricsFeature::InitialSyncStats cumulativeStats(vocbase().server().getFeature<ReplicationMetricsFeature>(), false);
```

**Review by graetzer — APPROVED** [review]

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 3]
```diff
@@ -959,13 +886,15 @@ Result DatabaseInitialSyncer::fetchCollectionDump(arangodb::LogicalCollection* c
...
+
+    cumulatedStats += stats;
 
     _config.progress.set(
         std::string("fetched leader collection dump for collection '") +
         coll->name() + "', type: " + typeString + ", id: " + leaderColl +
-        ", batch " + itoa(batch) + ", markers processed: " + itoa(markersProcessed) +
-        ", bytes received: " + itoa(bytesReceived) +
+        ", batch " + itoa(batch) + ", markers processed: " + itoa(cumulatedStats.numDumpDocuments) + 
+        ", bytes received: " + itoa(cumulatedStats.numDumpBytesReceived) +
```
Comment (danielhlarkin) [inline arangod/Replication/DatabaseInitialSyncer.cpp:897]:
```suggestion
        ", batch " + itoa(batch) + ", markers processed: " + itoa(cumulativeStats.numDumpDocuments) + 
        ", bytes received: " + itoa(cumulativeStats.numDumpBytesReceived) +
```

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 4]
```diff
@@ -977,11 +906,11 @@ Result DatabaseInitialSyncer::fetchCollectionDump(arangodb::LogicalCollection* c
...
-          ", markers processed: " + itoa(markersProcessed) +
-          ", bytes received: " + itoa(bytesReceived) +
-          ", dump requests: " + std::to_string(stats.numDumpRequests) +
-          ", waited for dump: " + std::to_string(stats.waitedForDump) + " s" +
-          ", apply time: " + std::to_string(stats.waitedForApply) + " s" +
+          ", markers processed: " + itoa(cumulatedStats.numDumpDocuments) + 
+          ", bytes received: " + itoa(cumulatedStats.numDumpBytesReceived) +
+          ", dump requests: " + std::to_string(cumulatedStats.numDumpRequests) +
+          ", waited for dump: " + std::to_string(cumulatedStats.waitedForDump) + " s" +
+          ", apply time: " + std::to_string(cumulatedStats.waitedForDumpApply) + " s" +
```
Comment (danielhlarkin) [inline arangod/Replication/DatabaseInitialSyncer.cpp:913]:
```suggestion
          ", markers processed: " + itoa(cumulativeStats.numDumpDocuments) + 
          ", bytes received: " + itoa(cumulativeStats.numDumpBytesReceived) +
          ", dump requests: " + std::to_string(cumulativeStats.numDumpRequests) +
          ", waited for dump: " + std::to_string(cumulativeStats.waitedForDump) + " s" +
          ", apply time: " + std::to_string(cumulativeStats.waitedForDumpApply) + " s" +
```

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 5]
```diff
@@ -991,6 +920,10 @@ Result DatabaseInitialSyncer::fetchCollectionDump(arangodb::LogicalCollection* c
     if (isAborted()) {
       return Result(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
     }
+
+    // update the global metrics so the changes become visible early
+    stats.publish();
+    // keep the cumulatedStats intact here!
```
Comment (danielhlarkin) [inline arangod/Replication/DatabaseInitialSyncer.cpp:926]:
```suggestion
    // keep the cumulativeStats intact here!
```

**Pull request merged**